src/test/scala/com/microsoft/hyperspace/index/IndexManagerTests.scala

```diff
@@ -399,13 +401,49 @@ class IndexManagerTests extends HyperspaceSuite with SQLHelper {
           case Seq(
               OptimizeActionEvent(_, _, "Operation started."),
               OptimizeActionEvent(_, _, msg)) =>
-            assert(msg.contains("Optimize aborted as no index files smaller than 1 found."))
+            assert(
+              msg.contains(
+                "Optimize aborted as no optimizable index files smaller than 1 found."))
           case _ => fail()
         }
       }
     }
   }
 
+  test("Verify optimize is no-op if each bucket has a single index file.") {
+    withTempPathAsString { testPath =>
+      val indexConfig = IndexConfig(s"index", Seq("RGUID"), Seq("imprs"))
+      import spark.implicits._
+      SampleData.testData
+        .toDF("Date", "RGUID", "Query", "imprs", "clicks")
+        .limit(10)
+        .write
+        .parquet(testPath)
+      val df = spark.read.parquet(testPath)
+      hyperspace.createIndex(df, indexConfig)
+
+      // Check if latest log file is updated with newly created index files.
+      val latestId = logManager(systemPath, indexConfig.indexName).getLatestId().get
+
+      MockEventLogger.reset()
+      hyperspace.optimizeIndex(indexConfig.indexName)
+
+      // Check that no new log files were created in this operation.
+      assert(latestId == logManager(systemPath, indexConfig.indexName).getLatestId().get)
+
+      // Check emitted events.
+      MockEventLogger.emittedEvents match {
+        case Seq(
+            OptimizeActionEvent(_, _, "Operation started."),
+            OptimizeActionEvent(_, _, msg)) =>
+          assert(
+            msg.contains(
+              "Optimize aborted as no optimizable index files smaller than 268435456 found."))
```

Comment:
Can we check all the index files are less than this number to be complete?

Reply:
Instead of this, added the check for the number of files per bucket.

Reply:
Shouldn't we check the size as well? Otherwise, `possibleCandidates` can be empty?

Reply:
I thought the size check was for the result of optimizeIndex. I added size check before optimizeIndex. Thanks!